### PR TITLE
Update short cut pin state properly on Windows

### PR DESCRIPTION
**Pull request by simonhong**
fix https://github.com/brave/brave-browser/issues/26301

Polling the pinned state after asking to Windows because it's actually pinned
when user allowed via os notification.
We'll check 10 times with 2s interval after asking to Windows.
If user doesn't react or disallowed within that time, we doesn't change the state.

Resolves

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

1. Launch brave and load brave://settings/getStarted
2. Click `Pin` button and click allow from os notification
3. Check Setting is changed to pinned state

### CODE REVIEW

File: browser/ui/webui/settings/pin_shortcut_handler.cc [hunk 1]
```diff
@@ -38,20 +38,63 @@ void PinShortcutHandler::HandleCheckShortcutPinState(
...
+#if BUILDFLAG(IS_WIN)
+  if (from_timer) {
+    pin_state_check_count_--;
+    // Clear if pinned or we did all try.
+    if (pinned || pin_state_check_count_ == 0) {
+      pin_state_check_count_ = 0;
+      pin_state_check_timer_.reset();
+    } else {
+      // Check again.
+      pin_state_check_timer_->Reset();
```
Comment (sangwoo108) [inline browser/ui/webui/settings/pin_shortcut_handler.cc:92]: How about using RepeatingTimer to remove this line?

File: browser/ui/webui/settings/pin_shortcut_handler.cc [hunk 2]
```diff
@@ -38,20 +38,63 @@ void PinShortcutHandler::HandleCheckShortcutPinState(
     const base::Value::List& args) {
   AllowJavascript();
 
+  CheckShortcutPinState(false);
```
Comment (sangwoo108) [inline browser/ui/webui/settings/pin_shortcut_handler.cc:41]: nit: for readers, might worth having comment for the arg

```suggestion
  CheckShortcutPinState(/*from_timer*/ false);
```

File: browser/ui/webui/settings/pin_shortcut_handler.cc [hunk 3]
```diff
@@ -38,20 +38,63 @@ void PinShortcutHandler::HandleCheckShortcutPinState(
...
+void PinShortcutHandler::OnPinStateCheckTimerFired() {
+  CheckShortcutPinState(true);
+}
+#endif
+
 void PinShortcutHandler::OnPinShortcut(bool pinned) {
   DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
 
+#if BUILDFLAG(IS_WIN)
+  // On Windows, ignore |pinned| value as Windows asks to user via another
```
Comment (sangwoo108) [inline browser/ui/webui/settings/pin_shortcut_handler.cc:60]: I'm curious if we still need this when |!pinned|. Does Windows prompt a dialog when unpinning shortcut from taskbar?

File: browser/ui/webui/settings/pin_shortcut_handler.cc (hunk 1 above)
Comment (simonhong) [reply browser/ui/webui/settings/pin_shortcut_handler.cc:96]: I think `RetainingOneShotTimer` is more fit here.
When it's fired, we need to post async call to get pinned state.
And then start timer again after getting async call's result.
If `RepeatingTimer` is used, next one is fired before checking pinned state.

File: browser/ui/webui/settings/pin_shortcut_handler.cc (hunk 2 above)
Comment (simonhong) [reply browser/ui/webui/settings/pin_shortcut_handler.cc]: Added.

File: browser/ui/webui/settings/pin_shortcut_handler.h [hunk 4]
```diff
@@ -27,8 +31,16 @@ class PinShortcutHandler : public settings::SettingsPageUIHandler {
...
 
+  void CheckShortcutPinState(bool from_timer);
   void OnPinShortcut(bool pinned);
-  void OnCheckShortcutPinState(bool pinned);
+  void OnCheckShortcutPinState(bool from_timer, bool pinned);
+
+#if BUILDFLAG(IS_WIN)
+  void OnPinStateCheckTimerFired();
+
+  int pin_state_check_count_ = 0;
```
[the remark this answers is not in the document]
Comment (simonhong) [reply browser/ui/webui/settings/pin_shortcut_handler.h]: seems much better :)

File: browser/ui/webui/settings/pin_shortcut_handler.cc (hunk 1 above)
Comment (sangwoo108) [reply browser/ui/webui/settings/pin_shortcut_handler.cc:96]: Aha, you're right. Thanks for explaning.

**Review by sangwoo108 — APPROVED** [review]
LGTM👍

**Review by mkarolin — APPROVED** [review]
`strings`++

**Comment by MadhaviSeelam** [comment]
Verification `In progress` using
```
Brave | 1.67.53 Chromium: 124.0.6367.91 (Official Build) nightly (64-bit)
-- | --
Revision | 418087e12068ca8ff9a74f346b9aac5032cd10de
OS | Windows 11 Version 23H2 (Build 22631.3527)

```

Case 1: New Install
<details>

1. Installed 1.67.53
2. Launch brave 
3. confirmed `Would like to Pin...` notification is shown
4. opned `brave://settings/getStarted` in a new tab
5. Clicked Pin button
6. confirmed a second notification is shown
7. clicked allow from os notification
8. Check Setting is changed to pinned state



step 3 | step 4 | step 6 | step 7 | step 8
---- | ----- | ---- | ----- | -----
image|image|image|image|image

**Pull request merged**